File: discover.py
#!/usr/bin/env python3
"""Discover git repos that need committing/pushing, using coding-agent records of
recently-active projects as the index — no filesystem scanning.

Pipeline:
  1. Collect candidate paths from every supported agent (union, deduped).
  2. Resolve each to its git toplevel (drops non-repos, parents, dups).
  3. Classify with `git status --porcelain=v2 --branch` + remote/upstream probing.
  4. Filter to dirty-or-unpushed and bucket: ready / commit_only / blocked.

The script reads each agent's config directly, so it sees ALL agents' recent
projects regardless of which agent is running it.

Adding a new agent = one function returning a list of candidate paths, registered
in SOURCES. Each source must fail closed (return [] on any error / when absent).

Outputs JSON on stdout. Use --pretty for a human-readable summary instead.
"""
from __future__ import annotations

import glob
import json
import os
import re
import subprocess
import sys
from urllib.parse import unquote, urlparse

HOME = os.path.expanduser("~")
IS_MAC = sys.platform == "darwin"


def run_git(repo: str, *args: str) -> tuple[int, str]:
    """Run a git command in `repo`, returning (returncode, stdout-stripped)."""
    try:
        p = subprocess.run(
            ["git", "-C", repo, *args],
            capture_output=True, text=True, timeout=20,
        )
        return p.returncode, p.stdout.strip()
    except Exception:
        return 1, ""


# --------------------------------------------------------------------------- #
# 1. Candidate sources (one per coding agent) — each fails closed.
# --------------------------------------------------------------------------- #
def src_claude_code() -> list[str]:
    """Claude Code — `~/.claude.json` -> .projects keys (literal absolute paths)."""
    path = os.path.join(HOME, ".claude.json")
    try:
        with open(path) as f:
            return list(json.load(f).get("projects", {}).keys())
    except Exception:
        return []


def src_codex() -> list[str]:
    """Codex — `~/.codex/config.toml` [projects."…"] headers, plus session cwd."""
    out: list[str] = []
    # Trusted-projects table in config.toml.
    cfg = os.path.join(HOME, ".codex", "config.toml")
    try:
        with open(cfg) as f:
            for line in f:
                m = re.match(r'^\[projects\."(.+)"\]\s*$', line)
                if m:
                    out.append(m.group(1))
    except Exception:
        pass
    # Session logs carry the actual cwd (complete record).
    root = os.path.join(HOME, ".codex", "sessions")
    try:
        files = glob.glob(os.path.join(root, "**", "*.jsonl"), recursive=True)
        files.sort(key=lambda p: os.path.getmtime(p), reverse=True)
        for fp in files[:800]:
            try:
                with open(fp) as f:
                    for _ in range(5):  # session_meta sits at the top of the file
                        line = f.readline()
                        if not line:
                            break
                        d = json.loads(line)
                        if not isinstance(d, dict):
                            continue
                        cwd = d.get("cwd") or (
                            d.get("payload", {}).get("cwd")
                            if isinstance(d.get("payload"), dict) else None
                        )
                        if cwd:
                            out.append(cwd)
                            break
            except Exception:
                continue
    except Exception:
        pass
    return out


def src_cursor() -> list[str]:
    """Cursor — VS Code-style state.vscdb -> history.recentlyOpenedPathsList."""
    import sqlite3
    if IS_MAC:
        db = os.path.join(HOME, "Library", "Application Support", "Cursor",
                          "User", "globalStorage", "state.vscdb")
    else:
        db = os.path.join(HOME, ".config", "Cursor",
                          "User", "globalStorage", "state.vscdb")
    if not os.path.isfile(db):
        return []
    out: list[str] = []
    try:
        con = sqlite3.connect(f"file:{db}?mode=ro", uri=True)
        try:
            row = con.execute(
                "SELECT value FROM ItemTable WHERE key=?",
                ("history.recentlyOpenedPathsList",),
            ).fetchone()
        finally:
            con.close()
        if row:
            for e in json.loads(row[0]).get("entries", []):
                u = e.get("folderUri") or e.get("fileUri")
                if isinstance(u, str) and u.startswith("file://"):
                    out.append(unquote(urlparse(u).path))
    except Exception:
        return []
    return out


# name -> collector. Add an agent here to extend discovery.
SOURCES = {
    "claude-code": src_claude_code,
    "codex": src_codex,
    "cursor": src_cursor,
}


def collect_candidates() -> tuple[list[str], dict[str, int]]:
    seen: set[str] = set()
    out: list[str] = []
    per_source: dict[str, int] = {}
    for name, fn in SOURCES.items():
        try:
            paths = fn() or []
        except Exception:
            paths = []
        per_source[name] = len(paths)
        for p in paths:
            if not p:
                continue
            p = os.path.expanduser(p.rstrip("/"))
            if p not in seen and os.path.isdir(p):
                seen.add(p)
                out.append(p)
    return out, per_source


# --------------------------------------------------------------------------- #
# 2. Resolve to git toplevel + dedup
# --------------------------------------------------------------------------- #
def resolve_repos(paths: list[str]) -> list[str]:
    roots, seen = [], set()
    for p in paths:
        rc, top = run_git(p, "rev-parse", "--show-toplevel")
        if rc == 0 and top and top not in seen:
            seen.add(top)
            roots.append(top)
    return roots


# --------------------------------------------------------------------------- #
# 3. Classify
# --------------------------------------------------------------------------- #
def classify(root: str) -> dict:
    rc, out = run_git(root, "status", "--porcelain=v2", "--branch")
    branch = "(unknown)"
    has_upstream = detached = False
    ahead = behind = dirty_count = 0

    for line in out.splitlines():
        if line.startswith("# branch.head "):
            branch = line[len("# branch.head "):]
            detached = branch == "(detached)"
        elif line.startswith("# branch.upstream "):
            has_upstream = True
        elif line.startswith("# branch.ab "):
            m = re.search(r"\+(\d+)\s+-(\d+)", line)
            if m:
                ahead, behind = int(m.group(1)), int(m.group(2))
        elif line and not line.startswith("#"):
            dirty_count += 1

    _, remotes = run_git(root, "remote")
    has_remote = bool(remotes.strip())

    # No upstream but a remote exists: probe origin/<branch> for implicit state.
    branch_on_remote = None
    if not has_upstream and has_remote and not detached:
        rc_v, _ = run_git(root, "rev-parse", "--verify", "--quiet", f"origin/{branch}")
        if rc_v == 0:
            branch_on_remote = True
            rc_c, counts = run_git(
                root, "rev-list", "--left-right", "--count", f"origin/{branch}...HEAD"
            )
            if rc_c == 0 and "\t" in counts:
                b, a = counts.split("\t")
                behind, ahead = int(b), int(a)
        else:
            branch_on_remote = False  # local branch never pushed

    dirty = dirty_count > 0
    needs_action = dirty or ahead > 0 or (has_remote and branch_on_remote is False)

    if not needs_action:
        bucket = "clean"
    elif behind > 0:
        bucket = "blocked"          # behind remote: commit local, don't auto-push
    elif not has_remote or (not has_upstream and branch_on_remote is False):
        bucket = "commit_only"      # nothing to push to / brand-new branch
    else:
        bucket = "ready"            # safe to commit + push

    return {
        "root": root, "branch": branch, "detached": detached,
        "dirty": dirty, "dirty_count": dirty_count,
        "ahead": ahead, "behind": behind,
        "has_remote": has_remote, "has_upstream": has_upstream,
        "branch_on_remote": branch_on_remote,
        "needs_action": needs_action, "bucket": bucket,
    }


def main() -> int:
    pretty = "--pretty" in sys.argv
    candidates, per_source = collect_candidates()
    roots = resolve_repos(candidates)
    repos = [classify(r) for r in roots]
    actionable = [r for r in repos if r["bucket"] != "clean"]

    buckets = {"ready": [], "commit_only": [], "blocked": []}
    for r in actionable:
        buckets[r["bucket"]].append(r)

    result = {
        "sources": per_source,
        "candidate_count": len(candidates),
        "repo_count": len(roots),
        "actionable_count": len(actionable),
        "clean_count": len(repos) - len(actionable),
        "buckets": buckets,
    }

    if not pretty:
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return 0

    labels = {
        "ready": "✅ READY (commit + push)",
        "commit_only": "📦 COMMIT-ONLY (no remote/upstream)",
        "blocked": "⚠️  BLOCKED (behind remote — won't auto-push)",
    }
    srcs = ", ".join(f"{k}={v}" for k, v in per_source.items())
    print(f"Sources: {srcs}")
    print(f"Scanned {len(candidates)} candidate paths -> {len(roots)} repos "
          f"({result['clean_count']} clean, {len(actionable)} need action)\n")
    for key in ("ready", "commit_only", "blocked"):
        items = buckets[key]
        if not items:
            continue
        print(labels[key])
        for r in items:
            bits = []
            if r["dirty"]:
                bits.append(f"{r['dirty_count']} changed")
            if r["ahead"]:
                bits.append(f"ahead {r['ahead']}")
            if r["behind"]:
                bits.append(f"behind {r['behind']}")
            if not r["has_remote"]:
                bits.append("no remote")
            elif r["branch_on_remote"] is False:
                bits.append("branch not on remote")
            print(f"  {r['root']}  [{r['branch']}]  ({', '.join(bits) or 'n/a'})")
        print()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: discover.ts
#!/usr/bin/env bun
/**
 * Discover git repos that need committing/pushing, using coding-agent records of
 * recently-active projects as the index — no filesystem scanning.
 *
 * Pipeline:
 *   1. Collect candidate paths from every supported agent (union, deduped).
 *   2. Resolve each to its git toplevel (drops non-repos, parents, dups).
 *   3. Classify with `git status --porcelain=v2 --branch` + remote/upstream probing.
 *   4. Filter to dirty-or-unpushed and bucket: ready / commit_only / blocked.
 *
 * The script reads each agent's config directly, so it sees ALL agents' recent
 * projects regardless of which agent is running it.
 *
 * Adding a new agent = one collector returning candidate paths, registered in
 * SOURCES. Each source must fail closed (return [] on any error / when absent).
 *
 * Run with Bun:  bun discover.ts [--pretty]
 *   (no flags) → JSON on stdout
 *   --pretty   → human-readable summary
 *
 * Zero dependencies: uses Bun built-ins (Bun.spawnSync, Bun.Glob, bun:sqlite).
 */
import { existsSync, statSync, readFileSync, openSync, readSync, closeSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { Glob } from "bun";
import { Database } from "bun:sqlite";

const HOME = homedir();
const IS_MAC = process.platform === "darwin";

function runGit(repo: string, ...args: string[]): { code: number; out: string } {
  try {
    const p = Bun.spawnSync(["git", "-C", repo, ...args], {
      stdout: "pipe",
      stderr: "pipe",
      timeout: 20_000,
    });
    return { code: p.exitCode, out: new TextDecoder().decode(p.stdout).trim() };
  } catch {
    return { code: 1, out: "" };
  }
}

/** Read only the head of a (possibly large) file, without loading the whole thing. */
function readHead(path: string, bytes = 65_536): string {
  const fd = openSync(path, "r");
  try {
    const buf = Buffer.alloc(bytes);
    const n = readSync(fd, buf, 0, bytes, 0);
    return buf.subarray(0, n).toString("utf8");
  } finally {
    closeSync(fd);
  }
}

// --------------------------------------------------------------------------- //
// 1. Candidate sources (one per coding agent) — each fails closed.
// --------------------------------------------------------------------------- //
type Source = () => string[];

/** Claude Code — `~/.claude.json` -> .projects keys (literal absolute paths). */
function srcClaudeCode(): string[] {
  try {
    const data = JSON.parse(readFileSync(join(HOME, ".claude.json"), "utf8"));
    return Object.keys(data?.projects ?? {});
  } catch {
    return [];
  }
}

/** Codex — `~/.codex/config.toml` [projects."…"] headers, plus session cwd. */
function srcCodex(): string[] {
  const out: string[] = [];

  // Trusted-projects table in config.toml.
  try {
    const cfg = readFileSync(join(HOME, ".codex", "config.toml"), "utf8");
    for (const line of cfg.split("\n")) {
      const m = line.match(/^\[projects\."(.+)"\]\s*$/);
      if (m) out.push(m[1]);
    }
  } catch {
    /* absent — skip */
  }

  // Session logs carry the actual cwd (complete record).
  try {
    const root = join(HOME, ".codex", "sessions");
    if (existsSync(root)) {
      const files: { p: string; m: number }[] = [];
      for (const rel of new Glob("**/*.jsonl").scanSync(root)) {
        const full = join(root, rel);
        try {
          files.push({ p: full, m: statSync(full).mtimeMs });
        } catch {
          /* skip unreadable */
        }
      }
      files.sort((a, b) => b.m - a.m); // most recent first
      for (const { p } of files.slice(0, 800)) {
        try {
          // session_meta sits at the top of the file.
          for (const line of readHead(p).split("\n").slice(0, 5)) {
            if (!line.trim()) continue;
            let d: any;
            try {
              d = JSON.parse(line);
            } catch {
              continue;
            }
            const cwd = d?.cwd ?? (typeof d?.payload === "object" ? d.payload?.cwd : null);
            if (cwd) {
              out.push(cwd);
              break;
            }
          }
        } catch {
          /* skip unreadable */
        }
      }
    }
  } catch {
    /* skip */
  }

  return out;
}

/** Cursor — VS Code-style state.vscdb -> history.recentlyOpenedPathsList. */
function srcCursor(): string[] {
  const db = IS_MAC
    ? join(HOME, "Library", "Application Support", "Cursor", "User", "globalStorage", "state.vscdb")
    : join(HOME, ".config", "Cursor", "User", "globalStorage", "state.vscdb");
  if (!existsSync(db)) return [];

  const out: string[] = [];
  try {
    const sqlite = new Database(db, { readonly: true });
    try {
      const row = sqlite
        .query("SELECT value FROM ItemTable WHERE key = ?")
        .get("history.recentlyOpenedPathsList") as { value: string } | null;
      if (row?.value) {
        for (const e of JSON.parse(row.value)?.entries ?? []) {
          const u: unknown = e.folderUri ?? e.fileUri;
          if (typeof u === "string" && u.startsWith("file://")) {
            out.push(decodeURIComponent(new URL(u).pathname));
          }
        }
      }
    } finally {
      sqlite.close();
    }
  } catch {
    return [];
  }
  return out;
}

// name -> collector. Add an agent here to extend discovery.
const SOURCES: Record<string, Source> = {
  "claude-code": srcClaudeCode,
  codex: srcCodex,
  cursor: srcCursor,
};

function collectCandidates(): { candidates: string[]; perSource: Record<string, number> } {
  const seen = new Set<string>();
  const candidates: string[] = [];
  const perSource: Record<string, number> = {};

  for (const [name, fn] of Object.entries(SOURCES)) {
    let paths: string[] = [];
    try {
      paths = fn() ?? [];
    } catch {
      paths = [];
    }
    perSource[name] = paths.length;
    for (let p of paths) {
      if (!p) continue;
      if (p.startsWith("~")) p = join(HOME, p.slice(1));
      p = p.replace(/\/+$/, "");
      try {
        if (!seen.has(p) && existsSync(p) && statSync(p).isDirectory()) {
          seen.add(p);
          candidates.push(p);
        }
      } catch {
        /* skip */
      }
    }
  }
  return { candidates, perSource };
}

// --------------------------------------------------------------------------- //
// 2. Resolve to git toplevel + dedup
// --------------------------------------------------------------------------- //
function resolveRepos(paths: string[]): string[] {
  const roots: string[] = [];
  const seen = new Set<string>();
  for (const p of paths) {
    const { code, out } = runGit(p, "rev-parse", "--show-toplevel");
    if (code === 0 && out && !seen.has(out)) {
      seen.add(out);
      roots.push(out);
    }
  }
  return roots;
}

// --------------------------------------------------------------------------- //
// 3. Classify
// --------------------------------------------------------------------------- //
type Bucket = "clean" | "ready" | "commit_only" | "blocked";

interface RepoStatus {
  root: string;
  branch: string;
  detached: boolean;
  dirty: boolean;
  dirtyCount: number;
  ahead: number;
  behind: number;
  hasRemote: boolean;
  hasUpstream: boolean;
  branchOnRemote: boolean | null;
  needsAction: boolean;
  bucket: Bucket;
}

function classify(root: string): RepoStatus {
  const { out } = runGit(root, "status", "--porcelain=v2", "--branch");
  let branch = "(unknown)";
  let hasUpstream = false;
  let detached = false;
  let ahead = 0;
  let behind = 0;
  let dirtyCount = 0;

  for (const line of out.split("\n")) {
    if (line.startsWith("# branch.head ")) {
      branch = line.slice("# branch.head ".length);
      detached = branch === "(detached)";
    } else if (line.startsWith("# branch.upstream ")) {
      hasUpstream = true;
    } else if (line.startsWith("# branch.ab ")) {
      const m = line.match(/\+(\d+)\s+-(\d+)/);
      if (m) {
        ahead = Number(m[1]);
        behind = Number(m[2]);
      }
    } else if (line && !line.startsWith("#")) {
      dirtyCount++;
    }
  }

  const hasRemote = runGit(root, "remote").out.trim().length > 0;

  // No upstream but a remote exists: probe origin/<branch> for implicit state.
  let branchOnRemote: boolean | null = null;
  if (!hasUpstream && hasRemote && !detached) {
    const v = runGit(root, "rev-parse", "--verify", "--quiet", `origin/${branch}`);
    if (v.code === 0) {
      branchOnRemote = true;
      const c = runGit(root, "rev-list", "--left-right", "--count", `origin/${branch}...HEAD`);
      if (c.code === 0 && c.out.includes("\t")) {
        const [b, a] = c.out.split("\t");
        behind = Number(b);
        ahead = Number(a);
      }
    } else {
      branchOnRemote = false; // local branch never pushed
    }
  }

  const dirty = dirtyCount > 0;
  const needsAction = dirty || ahead > 0 || (hasRemote && branchOnRemote === false);

  let bucket: Bucket;
  if (!needsAction) bucket = "clean";
  else if (behind > 0) bucket = "blocked"; // behind remote: commit local, don't auto-push
  else if (!hasRemote || (!hasUpstream && branchOnRemote === false)) bucket = "commit_only";
  else bucket = "ready";

  return {
    root, branch, detached, dirty, dirtyCount,
    ahead, behind, hasRemote, hasUpstream, branchOnRemote, needsAction, bucket,
  };
}

// --------------------------------------------------------------------------- //
// Main
// --------------------------------------------------------------------------- //
function main(): void {
  const pretty = process.argv.includes("--pretty");
  const { candidates, perSource } = collectCandidates();
  const roots = resolveRepos(candidates);
  const repos = roots.map(classify);
  const actionable = repos.filter((r) => r.bucket !== "clean");

  const buckets: Record<"ready" | "commit_only" | "blocked", RepoStatus[]> = {
    ready: [],
    commit_only: [],
    blocked: [],
  };
  for (const r of actionable) buckets[r.bucket as "ready" | "commit_only" | "blocked"].push(r);

  const result = {
    sources: perSource,
    candidate_count: candidates.length,
    repo_count: roots.length,
    actionable_count: actionable.length,
    clean_count: repos.length - actionable.length,
    buckets,
  };

  if (!pretty) {
    console.log(JSON.stringify(result, null, 2));
    return;
  }

  const labels: Record<string, string> = {
    ready: "✅ READY (commit + push)",
    commit_only: "📦 COMMIT-ONLY (no remote/upstream)",
    blocked: "⚠️  BLOCKED (behind remote — won't auto-push)",
  };
  const srcs = Object.entries(perSource).map(([k, v]) => `${k}=${v}`).join(", ");
  console.log(`Sources: ${srcs}`);
  console.log(
    `Scanned ${candidates.length} candidate paths -> ${roots.length} repos ` +
      `(${result.clean_count} clean, ${actionable.length} need action)\n`,
  );
  for (const key of ["ready", "commit_only", "blocked"] as const) {
    const items = buckets[key];
    if (!items.length) continue;
    console.log(labels[key]);
    for (const r of items) {
      const bits: string[] = [];
      if (r.dirty) bits.push(`${r.dirtyCount} changed`);
      if (r.ahead) bits.push(`ahead ${r.ahead}`);
      if (r.behind) bits.push(`behind ${r.behind}`);
      if (!r.hasRemote) bits.push("no remote");
      else if (r.branchOnRemote === false) bits.push("branch not on remote");
      console.log(`  ${r.root}  [${r.branch}]  (${bits.join(", ") || "n/a"})`);
    }
    console.log();
  }
}

main();
